File: core/src/main/java/com/netflix/iceberg/TableMetadata.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.netflix.iceberg;

import com.google.common.base.Objects;
import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.Iterables;
import com.google.common.collect.Lists;
import com.google.common.collect.Maps;
import com.google.common.collect.Sets;
import com.netflix.iceberg.exceptions.ValidationException;
import com.netflix.iceberg.io.InputFile;
import com.netflix.iceberg.types.TypeUtil;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Predicate;

/**
 * Metadata for a table.
 */
public class TableMetadata {
  static final int TABLE_FORMAT_VERSION = 1;

  public static TableMetadata newTableMetadata(TableOperations ops,
                                               Schema schema,
                                               PartitionSpec spec,
                                               String location) {
    return newTableMetadata(ops, schema, spec, location, ImmutableMap.of());
  }

  public static TableMetadata newTableMetadata(TableOperations ops,
                                               Schema schema,
                                               PartitionSpec spec,
                                               String location,
                                               Map<String, String> properties) {
    // reassign all column ids to ensure consistency
    AtomicInteger lastColumnId = new AtomicInteger(0);
    Schema freshSchema = TypeUtil.assignFreshIds(schema, lastColumnId::incrementAndGet);

    // rebuild the partition spec using the new column ids
    PartitionSpec.Builder specBuilder = PartitionSpec.builderFor(freshSchema);
    for (PartitionField field : spec.fields()) {
      // look up the name of the source field in the old schema to get the new schema's id
      String sourceName = schema.findColumnName(field.sourceId());
      specBuilder.add(
          freshSchema.findField(sourceName).fieldId(),
          field.name(),
          field.transform().toString());
    }
    PartitionSpec freshSpec = specBuilder.build();

    return new TableMetadata(ops, null, location,
        System.currentTimeMillis(),
        lastColumnId.get(), freshSchema, 0, ImmutableMap.of(0, freshSpec),
        ImmutableMap.copyOf(properties), -1, ImmutableList.of(), ImmutableList.of());
  }

  public static class SnapshotLogEntry {
    private final long timestampMillis;
    private final long snapshotId;

    SnapshotLogEntry(long timestampMillis, long snapshotId) {
      this.timestampMillis = timestampMillis;
      this.snapshotId = snapshotId;
    }

    public long timestampMillis() {
      return timestampMillis;
    }

    public long snapshotId() {
      return snapshotId;
    }

    @Override
    public boolean equals(Object other) {
      if (this == other) {
        return true;
      }
      if (other == null || getClass() != other.getClass()) {
        return false;
      }
      SnapshotLogEntry that = (SnapshotLogEntry) other;
      return timestampMillis == that.timestampMillis && snapshotId == that.snapshotId;
    }

    @Override
    public int hashCode() {
      return Objects.hashCode(timestampMillis, snapshotId);
    }

    @Override
    public String toString() {
      return Objects.toStringHelper(this)
          .add("timestampMillis", timestampMillis)
          .add("snapshotId", snapshotId)
          .toString();
    }
  }

  private final TableOperations ops;
  private final InputFile file;

  // stored metadata
  private final String location;
  private final long lastUpdatedMillis;
  private final int lastColumnId;
  private final Schema schema;
  private final int defaultSpecId;
  private final Map<Integer, PartitionSpec> specs;
  private final Map<String, String> properties;
  private final long currentSnapshotId;
  private final List<Snapshot> snapshots;
  private final Map<Long, Snapshot> snapshotsById;
  private final List<SnapshotLogEntry> snapshotLog;

  TableMetadata(TableOperations ops,
                InputFile file,
                String location,
                long lastUpdatedMillis,
                int lastColumnId,
                Schema schema,
                int defaultSpecId,
                Map<Integer, PartitionSpec> specs,
                Map<String, String> properties,
                long currentSnapshotId,
                List<Snapshot> snapshots,
                List<SnapshotLogEntry> snapshotLog) {
    this.ops = ops;
    this.file = file;
    this.location = location;
    this.lastUpdatedMillis = lastUpdatedMillis;
    this.lastColumnId = lastColumnId;
    this.schema = schema;
    this.specs = specs;
    this.defaultSpecId = defaultSpecId;
    this.properties = properties;
    this.currentSnapshotId = currentSnapshotId;
    this.snapshots = snapshots;
    this.snapshotLog = snapshotLog;

    ImmutableMap.Builder<Long, Snapshot> builder = ImmutableMap.builder();
    for (Snapshot version : snapshots) {
      builder.put(version.snapshotId(), version);
    }
    this.snapshotsById = builder.build();

    SnapshotLogEntry last = null;
    for (SnapshotLogEntry logEntry : snapshotLog) {
      if (last != null) {
        Preconditions.checkArgument(
            (logEntry.timestampMillis() - last.timestampMillis()) >= 0,
            "[BUG] Expected sorted snapshot log entries.");
      }
      last = logEntry;
    }

    Preconditions.checkArgument(
        currentSnapshotId < 0 || snapshotsById.containsKey(currentSnapshotId),
        "Invalid table metadata: Cannot find current version");
  }

  public InputFile file() {
    return file;
  }

  public long lastUpdatedMillis() {
    return lastUpdatedMillis;
  }

  public int lastColumnId() {
    return lastColumnId;
  }

  public Schema schema() {
    return schema;
  }

  public PartitionSpec spec() {
    return specs.get(defaultSpecId);
  }

  public int defaultSpecId() {
    return defaultSpecId;
  }

  public PartitionSpec spec(int id) {
    return specs.get(id);
  }

  public Map<Integer, PartitionSpec> specs() {
    return specs;
  }

  public String location() {
    return location;
  }

  public Map<String, String> properties() {
    return properties;
  }

  public int propertyAsInt(String property, int defaultValue) {
    String value = properties.get(property);
    if (value != null) {
      return Integer.parseInt(properties.get(property));
    }
    return defaultValue;
  }

  public long propertyAsLong(String property, long defaultValue) {
    String value = properties.get(property);
    if (value != null) {
      return Long.parseLong(properties.get(property));
    }
    return defaultValue;
  }

  public Snapshot snapshot(long snapshotId) {
    return snapshotsById.get(snapshotId);
  }

  public Snapshot currentSnapshot() {
    return snapshotsById.get(currentSnapshotId);
  }

  public List<Snapshot> snapshots() {
    return snapshots;
  }

  public List<SnapshotLogEntry> snapshotLog() {
    return snapshotLog;
  }

  public TableMetadata updateTableLocation(String newLocation) {
    return new TableMetadata(ops, null, newLocation,
        System.currentTimeMillis(), lastColumnId, schema, defaultSpecId, specs, properties,
        currentSnapshotId, snapshots, snapshotLog);
  }

  public TableMetadata updateSchema(Schema schema, int lastColumnId) {
    PartitionSpec.checkCompatibility(spec(), schema);
    return new TableMetadata(ops, null, location,
        System.currentTimeMillis(), lastColumnId, schema, defaultSpecId, specs, properties,
        currentSnapshotId, snapshots, snapshotLog);
  }

  public TableMetadata updatePartitionSpec(PartitionSpec spec) {
    PartitionSpec.checkCompatibility(spec, schema);

    // if the spec already exists, use the same ID. otherwise, use 1 more than the highest ID.
    int newDefaultSpecId = 0;
    for (Map.Entry<Integer, PartitionSpec> entry : specs.entrySet()) {
      if (spec.equals(entry.getValue())) {
        newDefaultSpecId = entry.getKey();
        break;
      } else if (newDefaultSpecId <= entry.getKey()) {
        newDefaultSpecId = entry.getKey() + 1;
      }
    }

    Preconditions.checkArgument(defaultSpecId != newDefaultSpecId,
        "Cannot set default partition spec to the current default");

    Map<Integer, PartitionSpec> newSpecs = Maps.newHashMap();
    newSpecs.putAll(specs);
    newSpecs.put(newDefaultSpecId, spec);

    return new TableMetadata(ops, null, location,
        System.currentTimeMillis(), lastColumnId, schema, newDefaultSpecId,
        ImmutableMap.copyOf(newSpecs), properties,
        currentSnapshotId, snapshots, snapshotLog);
  }

  public TableMetadata replaceCurrentSnapshot(Snapshot snapshot) {
    List<Snapshot> newSnapshots = ImmutableList.<Snapshot>builder()
        .addAll(snapshots)
        .add(snapshot)
        .build();
    List<SnapshotLogEntry> newSnapshotLog = ImmutableList.<SnapshotLogEntry>builder()
        .addAll(snapshotLog)
        .add(new SnapshotLogEntry(snapshot.timestampMillis(), snapshot.snapshotId()))
        .build();
    return new TableMetadata(ops, null, location,
        snapshot.timestampMillis(), lastColumnId, schema, defaultSpecId, specs, properties,
        snapshot.snapshotId(), newSnapshots, newSnapshotLog);
  }

  public TableMetadata removeSnapshotsIf(Predicate<Snapshot> removeIf) {
    List<Snapshot> filtered = Lists.newArrayListWithExpectedSize(snapshots.size());
    for (Snapshot snapshot : snapshots) {
      // keep the current snapshot and any snapshots that do not match the removeIf condition
      if (snapshot.snapshotId() == currentSnapshotId || !removeIf.test(snapshot)) {
        filtered.add(snapshot);
      }
    }

    // update the snapshot log
    Set<Long> validIds = Sets.newHashSet(Iterables.transform(filtered, Snapshot::snapshotId));
    List<SnapshotLogEntry> newSnapshotLog = Lists.newArrayList();
    for (SnapshotLogEntry logEntry : snapshotLog) {
      if (validIds.contains(logEntry.snapshotId())) {
        // copy the log entries that are still valid
        newSnapshotLog.add(logEntry);
      } else {
        // any invalid entry causes the history before it to be removed. otherwise, there could be
        // history gaps that cause time-travel queries to produce incorrect results. for example,
        // if history is [(t1, s1), (t2, s2), (t3, s3)] and s2 is removed, the history cannot be
        // [(t1, s1), (t3, s3)] because it appears that s3 was current during the time between t2
        // and t3 when in fact s2 was the current snapshot.
        newSnapshotLog.clear();
      }
    }

    return new TableMetadata(ops, null, location,
        System.currentTimeMillis(), lastColumnId, schema, defaultSpecId, specs, properties,
        currentSnapshotId, filtered, ImmutableList.copyOf(newSnapshotLog));
  }

  public TableMetadata rollbackTo(Snapshot snapshot) {
    ValidationException.check(snapshotsById.containsKey(snapshot.snapshotId()),
        "Cannot set current snapshot to unknown: %s", snapshot.snapshotId());

    long nowMillis = System.currentTimeMillis();
    List<SnapshotLogEntry> newSnapshotLog = ImmutableList.<SnapshotLogEntry>builder()
        .addAll(snapshotLog)
        .add(new SnapshotLogEntry(nowMillis, snapshot.snapshotId()))
        .build();

    return new TableMetadata(ops, null, location,
        nowMillis, lastColumnId, schema, defaultSpecId, specs, properties,
        snapshot.snapshotId(), snapshots, newSnapshotLog);
  }

  public TableMetadata replaceProperties(Map<String, String> newProperties) {
    ValidationException.check(newProperties != null, "Cannot set properties to null");
    return new TableMetadata(ops, null, location,
        System.currentTimeMillis(), lastColumnId, schema, defaultSpecId, specs, newProperties,
        currentSnapshotId, snapshots, snapshotLog);
  }

  public TableMetadata removeSnapshotLogEntries(Set<Long> snapshotIds) {
    List<SnapshotLogEntry> newSnapshotLog = Lists.newArrayList();
    for (SnapshotLogEntry logEntry : snapshotLog) {
      if (!snapshotIds.contains(logEntry.snapshotId())) {
        // copy the log entries that are still valid
        newSnapshotLog.add(logEntry);
      }
    }

    ValidationException.check(currentSnapshotId < 0 || // not set
            Iterables.getLast(newSnapshotLog).snapshotId() == currentSnapshotId,
        "Cannot set invalid snapshot log: latest entry is not the current snapshot");
    return new TableMetadata(ops, null, location,
        System.currentTimeMillis(), lastColumnId, schema, defaultSpecId, specs, properties,
        currentSnapshotId, snapshots, newSnapshotLog);
  }

  public TableMetadata buildReplacement(Schema schema, PartitionSpec partitionSpec,
                                        Map<String, String> properties) {
    AtomicInteger lastColumnId = new AtomicInteger(0);
    Schema freshSchema = TypeUtil.assignFreshIds(schema, lastColumnId::incrementAndGet);

    // rebuild the partition spec using the new column ids
    PartitionSpec.Builder specBuilder = PartitionSpec.builderFor(freshSchema);
    for (PartitionField field : partitionSpec.fields()) {
      // look up the name of the source field in the old schema to get the new schema's id
      String sourceName = schema.findColumnName(field.sourceId());
      specBuilder.add(
          freshSchema.findField(sourceName).fieldId(),
          field.name(),
          field.transform().toString());
    }
    PartitionSpec freshSpec = specBuilder.build();

    // if the spec already exists, use the same ID. otherwise, use 1 more than the highest ID.
    int specId = 0;
    for (Map.Entry<Integer, PartitionSpec> entry : specs.entrySet()) {
      if (freshSpec.equals(entry.getValue())) {
        specId = entry.getKey();
        break;
      } else if (specId <= entry.getKey()) {
        specId = entry.getKey() + 1;
      }
    }

    Map<Integer, PartitionSpec> newSpecs = Maps.newHashMap();
    newSpecs.putAll(specs);
    newSpecs.put(specId, freshSpec);

    Map<String, String> newProperties = Maps.newHashMap();
    newProperties.putAll(this.properties);
    newProperties.putAll(properties);

    return new TableMetadata(ops, null, location,
        System.currentTimeMillis(), lastColumnId.get(), freshSchema,
        specId, ImmutableMap.copyOf(newSpecs), ImmutableMap.copyOf(newProperties),
        -1, snapshots, ImmutableList.of());
  }
}

File: python/iceberg/core/table_metadata.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

import time
from typing import Optional

from iceberg.api import PartitionSpec, Schema
from iceberg.api.types import assign_fresh_ids
from iceberg.core.table_operations import TableOperations
from iceberg.core.util import AtomicInteger
from iceberg.exceptions import ValidationException


class TableMetadata(object):
    INITIAL_SPEC_ID = 0
    TABLE_FORMAT_VERSION = 1

    @staticmethod
    def new_table_metadata(ops: TableOperations, schema: Schema, spec: PartitionSpec, location: str,
                           properties: dict = None) -> "TableMetadata":
        last_column_id = AtomicInteger(0)
        fresh_schema = assign_fresh_ids(schema, last_column_id.increment_and_get)

        spec_builder = PartitionSpec.builder_for(fresh_schema)
        for field in spec.fields:
            src_name = schema.find_column_name(field.source_id)
            spec_builder.add(field.source_id,
                             fresh_schema.find_field(src_name).field_id,
                             field.name,
                             str(field.transform))

        fresh_spec = spec_builder.build()
        properties = properties if properties is not None else dict()

        return TableMetadata(ops, None, location,
                             int(time.time() * 1000),
                             last_column_id.get(), fresh_schema, TableMetadata.INITIAL_SPEC_ID, [fresh_spec],
                             properties, -1, list(), list())

    def __init__(self, ops, file, location, last_updated_millis,
                 last_column_id, schema, default_spec_id, specs, properties,
                 current_snapshot_id, snapshots, snapshot_log):
        self.ops = ops
        self._file = file
        self._location = location
        self.last_updated_millis = last_updated_millis
        self.last_column_id = last_column_id
        self.schema = schema
        self.default_spec_id = default_spec_id
        self.specs = specs
        self.properties = properties
        self.properties["provider"] = "ICEBERG"
        self.current_snapshot_id = current_snapshot_id
        self.snapshots = snapshots
        self.snapshot_log = snapshot_log

        self.snapshot_by_id = {version.snapshot_id: version for version in self.snapshots}
        self.specs_by_id = {spec.spec_id: spec for spec in self.specs}

        last = None
        for log_entry in snapshot_log:
            if last is not None:
                if not (log_entry.timestamp_millis - last.timestamp_millis > 0):
                    raise RuntimeError("[BUG] Expected sorted snapshot log entries.")
            last = log_entry

        if not (len(self.snapshot_by_id) == 0 or self.current_snapshot_id in self.snapshot_by_id):
            raise RuntimeError("Invalid table metadata: Cannot find current version")

    @property
    def location(self: "TableMetadata") -> str:
        return self._location

    @property
    def metadata_location(self: "TableMetadata") -> Optional[str]:
        return self._file.location() if self._file else None

    @property
    def spec(self):
        return self.specs_by_id[self.default_spec_id]

    def spec_id(self, spec_id):
        if spec_id is None:
            spec_id = self.default_spec_id

        return self.specs_by_id[spec_id]

    def property_as_int(self, property_name, default_value):
        return int(self.properties.get(property_name, default_value))

    def current_snapshot(self):
        return self.snapshot_by_id[self.current_snapshot_id]

    def snapshot(self, snapshot_id):
        return self.snapshot_by_id[snapshot_id]

    def update_metadata_location(self, new_location):
        return TableMetadata(self.ops, None, new_location,
                             int(time.time() * 1000), self.last_column_id, self.schema, self.spec, self.properties,
                             self.current_snapshot_id, self.snapshots, self.snapshot_log)

    def update_schema(self, schema, last_column_id):
        PartitionSpec.check_compatibility(self.spec, schema)
        return TableMetadata(self.ops, None, self.location,
                             int(time.time() * 1000), last_column_id, schema, self.spec, self.properties,
                             self.current_snapshot_id, self.snapshots, self.snapshot_log)

    def add_snapshot(self, snapshot):
        new_snapshots = self.snapshots + snapshot
        new_snapshot_log = self.snapshot_log + [SnapshotLogEntry(snapshot.timestamp_millis, snapshot.snapshot_id)]

        return TableMetadata(self.ops, None, self.location,
                             int(time.time() * 1000), self.last_column_id, self.schema, self.spec, self.properties,
                             self.current_snapshot_id, new_snapshots, new_snapshot_log)

    def add_staged_snapshot(self, snapshot):
        self.snapshots.append(snapshot)
        return TableMetadata(self.ops, None, self.location, snapshot.timestamp_millis,
                             self.last_column_id, self.schema, self.default_spec_id, self.specs,
                             self.current_snapshot_id, self.snapshots, self.snapshot_log)

    def replace_current_snapshot(self, snapshot):
        self.snapshots.append(snapshot)
        self.snapshot_log.append(SnapshotLogEntry(snapshot.timestamp_millis, snapshot.snapshot_id))

        return TableMetadata(self.ops, None, self.location, snapshot.timestamp_millis,
                             self.last_column_id, self.schema, self.default_spec_id, self.specs,
                             self.current_snapshot_id, self.snapshots, self.snapshot_log)

    def remove_snapshots_if(self, remove_if):
        filtered = list()

        for snapshot in self.snapshots:
            if snapshot.snapshot_id == self.current_snapshot_id or not remove_if(snapshot):
                filtered.append(snapshot)

        valid_ids = [snapshot.snapshot_id for snapshot in filtered]
        new_snapshot_log = list()
        for log_entry in self.snapshot_log:
            if log_entry.snapshot_id in valid_ids:
                new_snapshot_log.append(log_entry)
            else:
                new_snapshot_log.clear()

        return TableMetadata(self.ops, None, self.location,
                             int(time.time() * 1000), self.last_column_id, self.schema, self.spec, self.properties,
                             self.current_snapshot_id, filtered, new_snapshot_log)

    def rollback_to(self, snapshot):
        ValidationException.check(snapshot.snapshot_id not in self.snapshot_by_id,
                                  "Cannot set current snapshot to unknown: %s", (snapshot.snapshot_id,))

        now_millis = int(time.time() * 1000)
        new_snapshot_log = self.snapshot_log + [SnapshotLogEntry(now_millis, snapshot.snapshot_id)]

        return TableMetadata(self.ops, None, self.location,
                             now_millis, self.last_column_id, self.schema, self.spec, self.properties,
                             snapshot.snapshot_id, self.snapshots, new_snapshot_log)

    def replace_properties(self, new_properties):
        ValidationException.check(new_properties is not None, "Cannot set properties to null")

        return TableMetadata(self.ops, None, self.location,
                             int(time.time() * 1000), self.last_column_id, self.schema, self.spec, new_properties,
                             self.current_snapshot_id, self.snapshots, self.snapshot_log)

    def remove_snapshot_log_entries(self, snapshot_ids):
        new_snapshot_log = list()

        for entry in self.snapshot_log:
            if entry.snapshot_id not in snapshot_ids:
                new_snapshot_log.append(entry)

        check_snapshot = self.current_snapshot_id < 0 or new_snapshot_log[-1].snapshot_id == self.current_snapshot_id
        ValidationException.check(check_snapshot,
                                  "Cannot set invalid snapshot log: latest entry is not the current snapshot")

        return TableMetadata(self.ops, None, self.location,
                             int(time.time() * 1000), self.last_column_id, self.schema, self.spec, self.properties,
                             self.current_snapshot_id, self.snapshots, new_snapshot_log)


class SnapshotLogEntry(object):

    def __init__(self, timestamp_millis, snapshot_id):
        self.timestamp_millis = timestamp_millis
        self.snapshot_id = snapshot_id

    def __hash__(self):
        return hash(self.__key())

    def __key(self):
        return SnapshotLogEntry.__class__, self.snapshot_id, self.timestamp_millis

    def __eq__(self, other):
        if id(self) == id(other):
            return True
        if other is None or not isinstance(other, SnapshotLogEntry):
            return False

        return self.snapshot_id == other.snapshot_id and self.timestamp_millis == other.timestamp_millis

    def __ne__(self, other):
        return not self.__eq__()
